go/libkb/sig_chain.go

```diff
@@ -363,53 +364,85 @@ var hardcodedResets = map[keybase1.SigID]bool{
 
 // GetCurrentSubchain takes the given sigchain and walks backward until it
 // finds the start of the current subchain, returning all the links in the
-// subchain. A new subchain starts in one of three ways (from the perspective
-// of walking from oldest to newest):
-// 1) A link has a new eldest key, usually in the form of reporting no
-//    eldest_kid of its own and being signed by a KID that's not the previous
-//    eldest. Most resets so far take this form.
-// 2) A link of type "eldest", regardless of the KIDs involved. We want this to
-//    be how everything works in the future.
-// 3) One of a set of six hardcoded links that made it in back when we allowed
-//    repeating eldest keys without using the "eldest" link type.
-func (sc *SigChain) GetCurrentSubchain(eldest keybase1.KID) (links []*ChainLink, err error) {
-	if sc.chainLinks == nil {
-		return
-	}
-	l := len(sc.chainLinks)
-	lastGood := l
-	for i := l - 1; i >= 0; i-- {
-
-		if sc.chainLinks[i].IsStubbed() {
-			lastGood = i
-			if sc.chainLinks[i].IsEldest() {
-				break
-			}
-			continue
-		}
-
-		// Check that the eldest KID hasn't changed.
-		if sc.chainLinks[i].ToEldestKID().Equal(eldest) {
-			lastGood = i
-		} else {
-			break
-		}
-		// Also stop walking if the current link has type "eldest".
-		if sc.chainLinks[i].IsEldest() {
-			break
-		}
-		// Or if the link is one of our hardcoded six that reuse an eldest key ambiguously.
-		if hardcodedResets[sc.chainLinks[i].unpacked.sigID] {
-			break
+// subchain. See isSubchainStart for the details of the logic here.
+func (sc *SigChain) GetCurrentSubchain(eldest keybase1.KID) ([]*ChainLink, error) {
+	// Check for a totally empty chain (that is, a totally new account).
+	if len(sc.chainLinks) == 0 {
+		return nil, nil
+	}
+	// Confirm that the last link is not stubbed. This would prevent us from
+	// reading the eldest_kid, so the server should never do it.
+	lastLink := sc.chainLinks[len(sc.chainLinks)-1]
+	if lastLink.IsStubbed() {
+		return nil, errors.New("the last chain link is unexpectedly stubbed in GetCurrentSunchain")
+	}
+	// Check whether the eldest KID doesn't match the latest link. That means
+	// the account has just been reset, and so as with a new account, there is
+	// no current subchain.
+	if !sc.chainLinks[len(sc.chainLinks)-1].ToEldestKID().Equal(eldest) {
```

Comment:
This was on line 413 before. In the current version, `GetCurrentSubchain` is only responsible for checking the very last link against the expected eldest KID. Eldest KID changes prior to that in the chain are left to `isSubchainStart`, which is shared with cache loading, and which doesn't need an eldest KID from the Merkle tree.